File: sdks/go/aether/workflow/types_test.go
package workflow

import (
	"testing"
	"time"
)

func TestDuration_FromMillis(t *testing.T) {
	d := FromMillis(1500)
	if d.Milliseconds != 1500 {
		t.Errorf("expected 1500, got %d", d.Milliseconds)
	}
}

func TestDuration_FromSeconds(t *testing.T) {
	d := FromSeconds(2.5)
	if d.Milliseconds != 2500 {
		t.Errorf("expected 2500, got %d", d.Milliseconds)
	}
}

func TestDuration_FromMinutes(t *testing.T) {
	d := FromMinutes(1.5)
	if d.Milliseconds != 90000 {
		t.Errorf("expected 90000, got %d", d.Milliseconds)
	}
}

func TestDuration_FromHours(t *testing.T) {
	d := FromHours(2)
	if d.Milliseconds != 2*3600*1000 {
		t.Errorf("expected %d, got %d", 2*3600*1000, d.Milliseconds)
	}
}

func TestDuration_FromDays(t *testing.T) {
	d := FromDays(1)
	if d.Milliseconds != 24*3600*1000 {
		t.Errorf("expected %d, got %d", 24*3600*1000, d.Milliseconds)
	}
}

func TestDuration_FromTimeDuration(t *testing.T) {
	d := FromTimeDuration(5 * time.Second)
	if d.Milliseconds != 5000 {
		t.Errorf("expected 5000, got %d", d.Milliseconds)
	}
}

func TestDuration_ToTimeDuration(t *testing.T) {
	d := Duration{Milliseconds: 5000}
	td := d.ToTimeDuration()
	if td != 5*time.Second {
		t.Errorf("expected 5s, got %v", td)
	}
}

func TestDuration_TotalSeconds(t *testing.T) {
	d := Duration{Milliseconds: 2500}
	if d.TotalSeconds() != 2.5 {
		t.Errorf("expected 2.5, got %f", d.TotalSeconds())
	}
}

func TestDuration_Add(t *testing.T) {
	d1 := Duration{Milliseconds: 1000}
	d2 := Duration{Milliseconds: 500}
	result := d1.Add(d2)
	if result.Milliseconds != 1500 {
		t.Errorf("expected 1500, got %d", result.Milliseconds)
	}
}

func TestDuration_Sub(t *testing.T) {
	d1 := Duration{Milliseconds: 2000}
	d2 := Duration{Milliseconds: 500}
	result := d1.Sub(d2)
	if result.Milliseconds != 1500 {
		t.Errorf("expected 1500, got %d", result.Milliseconds)
	}
}

func TestSagaStatus_String(t *testing.T) {
	tests := []struct {
		s    SagaStatus
		want string
	}{
		{SagaStatusPending, "pending"},
		{SagaStatusRunning, "running"},
		{SagaStatusCompleted, "completed"},
		{SagaStatusCompensating, "compensating"},
		{SagaStatusCompensated, "compensated"},
		{SagaStatusFailed, "failed"},
		{SagaStatus(99), "unknown"},
	}
	for _, tt := range tests {
		if got := tt.s.String(); got != tt.want {
			t.Errorf("SagaStatus(%d).String() = %q, want %q", tt.s, got, tt.want)
		}
	}
}

func TestStepStatus_String(t *testing.T) {
	tests := []struct {
		s    StepStatus
		want string
	}{
		{StepStatusPending, "pending"},
		{StepStatusRunning, "running"},
		{StepStatusCompleted, "completed"},
		{StepStatusCompensating, "compensating"},
		{StepStatusCompensated, "compensated"},
		{StepStatusFailed, "failed"},
		{StepStatusSkipped, "skipped"},
		{StepStatus(99), "unknown"},
	}
	for _, tt := range tests {
		if got := tt.s.String(); got != tt.want {
			t.Errorf("StepStatus(%d).String() = %q, want %q", tt.s, got, tt.want)
		}
	}
}

func TestWorkflowStatus_String(t *testing.T) {
	tests := []struct {
		s    WorkflowStatus
		want string
	}{
		{WorkflowStatusCreated, "created"},
		{WorkflowStatusRunning, "running"},
		{WorkflowStatusSuspended, "suspended"},
		{WorkflowStatusCompleted, "completed"},
		{WorkflowStatusFailed, "failed"},
		{WorkflowStatusCancelled, "cancelled"},
		{WorkflowStatus(99), "unknown"},
	}
	for _, tt := range tests {
		if got := tt.s.String(); got != tt.want {
			t.Errorf("WorkflowStatus(%d).String() = %q, want %q", tt.s, got, tt.want)
		}
	}
}

func TestTransitionStatus_String(t *testing.T) {
	tests := []struct {
		s    TransitionStatus
		want string
	}{
		{TransitionStatusPending, "pending"},
		{TransitionStatusSuccess, "success"},
		{TransitionStatusFailed, "failed"},
		{TransitionStatusRolledBack, "rolled_back"},
		{TransitionStatus(99), "unknown"},
	}
	for _, tt := range tests {
		if got := tt.s.String(); got != tt.want {
			t.Errorf("TransitionStatus(%d).String() = %q, want %q", tt.s, got, tt.want)
		}
	}
}

func TestHumanTaskStatus_String(t *testing.T) {
	tests := []struct {
		s    HumanTaskStatus
		want string
	}{
		{HumanTaskStatusPending, "pending"},
		{HumanTaskStatusAssigned, "assigned"},
		{HumanTaskStatusInProgress, "in_progress"},
		{HumanTaskStatusCompleted, "completed"},
		{HumanTaskStatusRejected, "rejected"},
		{HumanTaskStatusTimeout, "timeout"},
		{HumanTaskStatusEscalated, "escalated"},
		{HumanTaskStatus(99), "unknown"},
	}
	for _, tt := range tests {
		if got := tt.s.String(); got != tt.want {
			t.Errorf("HumanTaskStatus(%d).String() = %q, want %q", tt.s, got, tt.want)
		}
	}
}

func TestRetryPolicy_String(t *testing.T) {
	tests := []struct {
		p    RetryPolicy
		want string
	}{
		{RetryPolicyNone, "none"},
		{RetryPolicyFixed, "fixed"},
		{RetryPolicyExponential, "exponential"},
		{RetryPolicyExponentialJitter, "exponential_jitter"},
		{RetryPolicy(99), "unknown"},
	}
	for _, tt := range tests {
		if got := tt.p.String(); got != tt.want {
			t.Errorf("RetryPolicy(%d).String() = %q, want %q", tt.p, got, tt.want)
		}
	}
}

func TestDefaultRetryConfig(t *testing.T) {
	cfg := DefaultRetryConfig()
	if cfg.MaxAttempts != 3 {
		t.Errorf("expected 3, got %d", cfg.MaxAttempts)
	}
	if cfg.Policy != RetryPolicyExponential {
		t.Errorf("expected exponential, got %v", cfg.Policy)
	}
	if cfg.InitialDelay.TotalSeconds() != 1.0 {
		t.Errorf("expected 1s, got %f", cfg.InitialDelay.TotalSeconds())
	}
}

func TestSagaContext_New(t *testing.T) {
	ctx := NewSagaContext("input-data")
	if ctx == nil {
		t.Fatal("expected non-nil context")
	}
	if ctx.SagaID == "" {
		t.Error("expected non-empty saga ID")
	}
	if ctx.Input != "input-data" {
		t.Errorf("expected 'input-data', got %v", ctx.Input)
	}
	if ctx.StartedAt == nil {
		t.Error("expected started at to be set")
	}
}

func TestSagaContext_State(t *testing.T) {
	ctx := NewSagaContext(nil)
	ctx.SetState("key1", "value1")
	ctx.SetState("key2", 42)

	if ctx.GetState("key1") != "value1" {
		t.Errorf("expected 'value1', got %v", ctx.GetState("key1"))
	}
	if ctx.GetState("key2") != 42 {
		t.Errorf("expected 42, got %v", ctx.GetState("key2"))
	}
	if ctx.GetStateDefault("missing", "default") != "default" {
		t.Errorf("expected 'default', got %v", ctx.GetStateDefault("missing", "default"))
	}
	if ctx.GetStateDefault("key1", "fallback") != "value1" {
		t.Errorf("expected 'value1', got %v", ctx.GetStateDefault("key1", "fallback"))
	}
}

func TestSagaContext_StepCompletion(t *testing.T) {
	ctx := NewSagaContext(nil)
	ctx.MarkStepCompleted("step1")
	ctx.MarkStepCompleted("step2")

	if !ctx.IsStepCompleted("step1") {
		t.Error("step1 should be completed")
	}
	if !ctx.IsStepCompleted("step2") {
		t.Error("step2 should be completed")
	}
	if ctx.IsStepCompleted("step3") {
		t.Error("step3 should not be completed")
	}

	ctx.MarkStepCompleted("step1")
	if len(ctx.CompletedSteps) != 2 {
		t.Errorf("duplicate completion should not add, got %d", len(ctx.CompletedSteps))
	}
}

func TestWorkflowContext_New(t *testing.T) {
	ctx := NewWorkflowContext("test-workflow", "input")
	if ctx == nil {
		t.Fatal("expected non-nil context")
	}
	if ctx.WorkflowType != "test-workflow" {
		t.Errorf("expected 'test-workflow', got %q", ctx.WorkflowType)
	}
	if ctx.Status != WorkflowStatusCreated {
		t.Errorf("expected created status, got %v", ctx.Status)
	}
}

func TestWorkflowContext_Variables(t *testing.T) {
	ctx := NewWorkflowContext("test", nil)
	ctx.SetVariable("key1", "value1")
	ctx.SetVariable("key2", 42)

	if ctx.GetVariable("key1") != "value1" {
		t.Errorf("expected 'value1', got %v", ctx.GetVariable("key1"))
	}
	if ctx.GetVariableDefault("missing", "default") != "default" {
		t.Errorf("expected 'default', got %v", ctx.GetVariableDefault("missing", "default"))
	}
}

func TestWorkflowContext_History(t *testing.T) {
	ctx := NewWorkflowContext("test", nil)
	ctx.AddHistoryEvent("start", map[string]interface{}{"key": "val"})
	ctx.AddHistoryEvent("complete", nil)

	if len(ctx.History) != 2 {
		t.Errorf("expected 2 events, got %d", len(ctx.History))
	}
	if ctx.History[0].Type != "start" {
		t.Errorf("expected 'start', got %q", ctx.History[0].Type)
	}
	if ctx.UpdatedAt == nil {
		t.Error("updated at should be set")
	}
}

func TestHumanTaskContext_New(t *testing.T) {
	ctx := NewHumanTaskContext("approval", "Approve Request")
	if ctx == nil {
		t.Fatal("expected non-nil context")
	}
	if ctx.Title != "Approve Request" {
		t.Errorf("expected 'Approve Request', got %q", ctx.Title)
	}
	if ctx.Priority != 5 {
		t.Errorf("expected priority 5, got %d", ctx.Priority)
	}
	if ctx.Status != HumanTaskStatusPending {
		t.Errorf("expected pending status, got %v", ctx.Status)
	}
}

func TestSagaError(t *testing.T) {
	err := NewSagaError("step1", nil)
	if err.Error() == "" {
		t.Error("error message should not be empty")
	}
	if !contains(err.Error(), "step1") {
		t.Errorf("error should contain step name, got %q", err.Error())
	}

	err2 := NewSagaError("step2", &SagaError{StepName: "inner"})
	if !contains(err2.Error(), "step2") {
		t.Errorf("error should contain step name, got %q", err2.Error())
	}
}

func TestCompensationError(t *testing.T) {
	err := NewCompensationError("step1", nil)
	if !contains(err.Error(), "step1") {
		t.Errorf("error should contain step name, got %q", err.Error())
	}
}

func TestWorkflowError(t *testing.T) {
	err := NewWorkflowError("something failed")
	if err.Error() != "something failed" {
		t.Errorf("expected 'something failed', got %q", err.Error())
	}
}

func TestInvalidTransitionError(t *testing.T) {
	err := NewInvalidTransitionError("start", "end", "wf-1")
	msg := err.Error()
	if !contains(msg, "start") || !contains(msg, "end") || !contains(msg, "wf-1") {
		t.Errorf("error should contain all parts, got %q", msg)
	}
}

func TestHumanTaskError(t *testing.T) {
	err := NewHumanTaskError("task-1", "assignment failed")
	if !contains(err.Error(), "task-1") || !contains(err.Error(), "assignment failed") {
		t.Errorf("error should contain parts, got %q", err.Error())
	}
}

func TestSagaResult_Fields(t *testing.T) {
	now := time.Now()
	sr := SagaResult{
		SagaID:     "saga-1",
		Status:     SagaStatusCompleted,
		Output:     "result",
		StartedAt:  &now,
		DurationMs: 5000,
	}
	if sr.SagaID != "saga-1" {
		t.Errorf("expected 'saga-1', got %q", sr.SagaID)
	}
}

func TestWorkflowResult_Fields(t *testing.T) {
	now := time.Now()
	wr := WorkflowResult{
		WorkflowID:   "wf-1",
		Status:       WorkflowStatusCompleted,
		Output:       "result",
		StartedAt:    &now,
		DurationMs:   3000,
		CurrentState: "done",
	}
	if wr.WorkflowID != "wf-1" {
		t.Errorf("expected 'wf-1', got %q", wr.WorkflowID)
	}
}

func TestTransitionResult_Fields(t *testing.T) {
	tr := TransitionResult{
		Success:   true,
		FromState: "start",
		ToState:   "end",
		Timestamp: time.Now(),
	}
	if !tr.Success {
		t.Error("expected success")
	}
}

func TestState_Fields(t *testing.T) {
	s := State{
		Name:      "initial",
		IsInitial: true,
		IsFinal:   false,
		Metadata:  map[string]interface{}{"key": "val"},
	}
	if !s.IsInitial {
		t.Error("should be initial")
	}
}

func TestTransition_Fields(t *testing.T) {
	tr := Transition{
		Name:      "start-to-process",
		FromState: "start",
		ToState:   "process",
		Metadata:  map[string]interface{}{"auto": true},
	}
	if tr.Name != "start-to-process" {
		t.Errorf("expected 'start-to-process', got %q", tr.Name)
	}
}

func TestSagaStep_Fields(t *testing.T) {
	now := time.Now()
	step := SagaStep{
		Name:        "create-order",
		Status:      StepStatusCompleted,
		Attempts:    1,
		StartedAt:   &now,
		CompletedAt: &now,
	}
	if step.Name != "create-order" {
		t.Errorf("expected 'create-order', got %q", step.Name)
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && containsStr(s, substr)
}

func containsStr(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

File: sdks/javascript/tests/workflow/types.test.ts
import {
  Duration,
  SagaStepStatus,
  SagaStatus,
  WorkflowStatus,
  HumanTaskStatus,
  StateMachineEventType,
  RetryPolicy,
  defaultRetryConfig,
  SagaContext,
  WorkflowContext,
  SagaError,
  SagaStepFailedError,
  SagaCompensationFailedError,
  WorkflowError,
  InvalidTransitionError,
  WorkflowSuspendedError,
  HumanTaskError,
  HumanTaskTimeoutError,
  HumanTaskNotAssignedError,
} from '../../src/workflow/types';

describe('Duration', () => {
  it('creates from milliseconds', () => {
    const d = Duration.milliseconds(1500);
    expect(d.toMilliseconds()).toBe(1500);
  });

  it('creates from seconds', () => {
    const d = Duration.seconds(30);
    expect(d.toMilliseconds()).toBe(30000);
    expect(d.toSeconds()).toBe(30);
  });

  it('creates from minutes', () => {
    const d = Duration.minutes(5);
    expect(d.toMilliseconds()).toBe(300000);
    expect(d.toMinutes()).toBe(5);
  });

  it('creates from hours', () => {
    const d = Duration.hours(2);
    expect(d.toMilliseconds()).toBe(7200000);
    expect(d.toHours()).toBe(2);
  });

  it('creates from days', () => {
    const d = Duration.days(1);
    expect(d.toMilliseconds()).toBe(86400000);
  });

  it('rounds fractional inputs', () => {
    const d = Duration.milliseconds(1.7);
    expect(d.toMilliseconds()).toBe(2);
  });

  it('adds two durations', () => {
    const a = Duration.seconds(10);
    const b = Duration.seconds(20);
    expect(a.add(b).toSeconds()).toBe(30);
  });

  it('subtracts durations', () => {
    const a = Duration.seconds(30);
    const b = Duration.seconds(10);
    expect(a.subtract(b).toSeconds()).toBe(20);
  });

  it('clamps subtraction to zero', () => {
    const a = Duration.seconds(5);
    const b = Duration.seconds(10);
    expect(a.subtract(b).toMilliseconds()).toBe(0);
  });

  it('converts between units correctly', () => {
    const d = Duration.hours(1);
    expect(d.toMinutes()).toBe(60);
    expect(d.toSeconds()).toBe(3600);
  });

  it('handles zero duration', () => {
    const d = Duration.milliseconds(0);
    expect(d.toMilliseconds()).toBe(0);
  });

  it('handles negative input by rounding toward zero', () => {
    const d = Duration.seconds(-1);
    expect(d.toMilliseconds()).toBe(-1000);
  });
});

describe('Enums', () => {
  it('SagaStepStatus has expected values', () => {
    expect(SagaStepStatus.Pending).toBe('pending');
    expect(SagaStepStatus.Running).toBe('running');
    expect(SagaStepStatus.Completed).toBe('completed');
    expect(SagaStepStatus.Compensating).toBe('compensating');
    expect(SagaStepStatus.Compensated).toBe('compensated');
    expect(SagaStepStatus.Failed).toBe('failed');
    expect(SagaStepStatus.Skipped).toBe('skipped');
  });

  it('SagaStatus has expected values', () => {
    expect(SagaStatus.Pending).toBe('pending');
    expect(SagaStatus.Running).toBe('running');
    expect(SagaStatus.Completed).toBe('completed');
    expect(SagaStatus.Compensating).toBe('compensating');
    expect(SagaStatus.Compensated).toBe('compensated');
    expect(SagaStatus.Failed).toBe('failed');
  });

  it('WorkflowStatus has expected values', () => {
    expect(WorkflowStatus.Created).toBe('created');
    expect(WorkflowStatus.Running).toBe('running');
    expect(WorkflowStatus.Suspended).toBe('suspended');
    expect(WorkflowStatus.Completed).toBe('completed');
    expect(WorkflowStatus.Failed).toBe('failed');
    expect(WorkflowStatus.Cancelled).toBe('cancelled');
  });

  it('HumanTaskStatus has expected values', () => {
    expect(HumanTaskStatus.Pending).toBe('pending');
    expect(HumanTaskStatus.Assigned).toBe('assigned');
    expect(HumanTaskStatus.InProgress).toBe('in-progress');
    expect(HumanTaskStatus.Completed).toBe('completed');
    expect(HumanTaskStatus.Rejected).toBe('rejected');
    expect(HumanTaskStatus.Timeout).toBe('timeout');
    expect(HumanTaskStatus.Escalated).toBe('escalated');
  });

  it('StateMachineEventType has expected values', () => {
    expect(StateMachineEventType.Enter).toBe('enter');
    expect(StateMachineEventType.Exit).toBe('exit');
    expect(StateMachineEventType.Transition).toBe('transition');
    expect(StateMachineEventType.GuardFailed).toBe('guard-failed');
    expect(StateMachineEventType.ActionFailed).toBe('action-failed');
  });

  it('RetryPolicy enum has expected values', () => {
    expect(RetryPolicy.None).toBe('none');
    expect(RetryPolicy.Fixed).toBe('fixed');
    expect(RetryPolicy.Exponential).toBe('exponential');
    expect(RetryPolicy.ExponentialJitter).toBe('exponential-jitter');
  });
});

describe('defaultRetryConfig', () => {
  it('returns sensible defaults', () => {
    const cfg = defaultRetryConfig();
    expect(cfg.maxAttempts).toBe(3);
    expect(cfg.policy).toBe(RetryPolicy.Exponential);
    expect(cfg.initialDelay.toSeconds()).toBe(1);
    expect(cfg.maxDelay.toSeconds()).toBe(60);
    expect(cfg.multiplier).toBe(2.0);
    expect(cfg.jitter).toBe(0.1);
  });
});

describe('SagaContext', () => {
  it('generates a sagaId if none provided', () => {
    const ctx = new SagaContext();
    expect(ctx.sagaId).toBeTruthy();
    expect(typeof ctx.sagaId).toBe('string');
  });

  it('uses provided sagaId', () => {
    const ctx = new SagaContext('my-id');
    expect(ctx.sagaId).toBe('my-id');
  });

  it('sets and gets state', () => {
    const ctx = new SagaContext();
    ctx.setState('key', 'value');
    expect(ctx.getState('key')).toBe('value');
  });

  it('returns defaultValue for missing state key', () => {
    const ctx = new SagaContext();
    expect(ctx.getState('missing', 'fallback')).toBe('fallback');
    expect(ctx.getState('missing')).toBeUndefined();
  });

  it('tracks completed steps', () => {
    const ctx = new SagaContext();
    ctx.markStepCompleted('step1');
    ctx.markStepCompleted('step2');
    expect(ctx.completedSteps).toEqual(['step1', 'step2']);
  });

  it('does not duplicate completed steps', () => {
    const ctx = new SagaContext();
    ctx.markStepCompleted('step1');
    ctx.markStepCompleted('step1');
    expect(ctx.completedSteps).toEqual(['step1']);
  });

  it('checks step completion', () => {
    const ctx = new SagaContext();
    ctx.markStepCompleted('done');
    expect(ctx.isStepCompleted('done')).toBe(true);
    expect(ctx.isStepCompleted('other')).toBe(false);
  });

  it('stores typed input', () => {
    const ctx = new SagaContext<{ orderId: string }>();
    ctx.input = { orderId: '123' };
    expect(ctx.input?.orderId).toBe('123');
  });
});

describe('WorkflowContext', () => {
  it('generates a workflowId if none provided', () => {
    const ctx = new WorkflowContext();
    expect(ctx.workflowId).toBeTruthy();
  });

  it('uses provided workflowId', () => {
    const ctx = new WorkflowContext('wf-1');
    expect(ctx.workflowId).toBe('wf-1');
  });

  it('defaults to Running status', () => {
    const ctx = new WorkflowContext();
    expect(ctx.status).toBe(WorkflowStatus.Running);
  });

  it('sets and gets variables', () => {
    const ctx = new WorkflowContext();
    ctx.setVariable('x', 42);
    expect(ctx.getVariable('x')).toBe(42);
    expect(ctx.getVariable('y', 0)).toBe(0);
  });

  it('records history events', () => {
    const ctx = new WorkflowContext();
    ctx.addHistoryEvent('test-event', { detail: true });
    expect(ctx.history).toHaveLength(1);
    expect(ctx.history[0].type).toBe('test-event');
    expect(ctx.history[0].detail).toBe(true);
  });

  it('adds timestamp to history events', () => {
    const ctx = new WorkflowContext();
    ctx.addHistoryEvent('ev');
    expect(ctx.history[0].timestamp).toBeTruthy();
  });
});

describe('Errors', () => {
  it('SagaError extends Error with correct name', () => {
    const err = new SagaError('saga broke');
    expect(err).toBeInstanceOf(Error);
    expect(err).toBeInstanceOf(SagaError);
    expect(err.name).toBe('SagaError');
    expect(err.message).toBe('saga broke');
  });

  it('SagaStepFailedError carries step name and cause', () => {
    const cause = new Error('boom');
    const err = new SagaStepFailedError('step1', cause);
    expect(err.stepName).toBe('step1');
    expect(err.cause).toBe(cause);
    expect(err.message).toContain('step1');
    expect(err.message).toContain('boom');
  });

  it('SagaStepFailedError handles unknown cause', () => {
    const err = new SagaStepFailedError('step2');
    expect(err.message).toContain('step2');
    expect(err.message).toContain('unknown error');
  });

  it('SagaCompensationFailedError carries step name and cause', () => {
    const err = new SagaCompensationFailedError('step1', new Error('fail'));
    expect(err.stepName).toBe('step1');
    expect(err.name).toBe('SagaCompensationFailedError');
  });

  it('WorkflowError extends AetherError', () => {
    const err = new WorkflowError('wf err');
    expect(err).toBeInstanceOf(Error);
    expect(err.name).toBe('WorkflowError');
  });

  it('InvalidTransitionError carries from/to states', () => {
    const err = new InvalidTransitionError('A', 'B', 'wf-1');
    expect(err.fromState).toBe('A');
    expect(err.toState).toBe('B');
    expect(err.workflowId).toBe('wf-1');
    expect(err.message).toContain('A');
    expect(err.message).toContain('B');
  });

  it('WorkflowSuspendedError carries workflowId', () => {
    const err = new WorkflowSuspendedError('wf-1', 'maintenance');
    expect(err.workflowId).toBe('wf-1');
    expect(err.message).toContain('maintenance');
  });

  it('HumanTaskError extends AetherError', () => {
    const err = new HumanTaskError('task err');
    expect(err).toBeInstanceOf(Error);
    expect(err.name).toBe('HumanTaskError');
  });

  it('HumanTaskTimeoutError carries taskId', () => {
    const err = new HumanTaskTimeoutError('task-1');
    expect(err.taskId).toBe('task-1');
    expect(err.name).toBe('HumanTaskTimeoutError');
  });

  it('HumanTaskNotAssignedError carries taskId', () => {
    const err = new HumanTaskNotAssignedError('task-2');
    expect(err.taskId).toBe('task-2');
    expect(err.name).toBe('HumanTaskNotAssignedError');
  });
});
